### PR TITLE
Improve catalog rule product indexing performance

### PR DESCRIPTION
**Pull request by rossmitchell**
### Description (*)

Updates the IndexBuilder class to call `fetchItem` on the Collection rather than `foreach`. This prevents the entire collection being loaded at once which significantly speeds up the indexing time

### Fixed Issues (if relevant)

1. magento/magento2#25889: Catalog Rule Product index time increases exponentially as new rules are added

### Manual testing scenarios (*)

1. Install Magento
2. Copy setup/performance-toolkit/profiles/ce/small.xml to a new file
3. Update the file to have 2,000 simple products, 2,000 categories, and 25,000 catalog price rules
4. Generate fixtures using the file
5. Run the Catalog Rule Product index
6. The index should complete in a few minutes rather than an hour

### Questions or comments

Alongside the change mentioned in the ticket, there are a couple of other places in the class that were calling `foreach` on the collection. I've applied the same update to them as well
 
I'm not sure of the best way to put together a test to cover these changes, the behaviour should be the same just faster. If you can give any guidance on what should be included I'll add in tests for this.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Review by aleron75 — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @aleron75, thank you for the review.
[ENGCOM-6437](https://jira.corp.magento.com/browse/ENGCOM-6437) has been created to process this Pull Request

**Comment by magento-engcom-team** [comment]
@rossmitchell thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed


 With 25,000 catalog price rules (with fix)
```
$ php bin/magento indexer:reindex catalogrule_rule
Catalog Rule Product index has been rebuilt successfully in 00:04:44
Product Price index has been rebuilt successfully in 00:00:02
Catalog Search index has been rebuilt successfully in 00:00:10
```

**Review by VladimirZaets — CHANGES_REQUESTED** [review]
Hi @rossmitchell. Thanks for collaboration. 
Due to [Magento Definition of Done](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing_dod.html) all code must be covered by tests. Please cover your fix by automated tests.

**Comment by rossmitchell** [comment]
Hi @VladimirZaets 

As I mentioned in the opening ticket, I'm not sure how to put together a test for these changes. They shouldn't change any existing behaviour, just they way that records are fetched from the database.

If you can give me some guidance over what sort of tests should be written to cover the changes, I'll add them and update the PR. Thanks

**Comment by lenaorobei** [comment]
I run Performance build on this branch and got `Create category (Admin Category Management): 10.7% (mainline = 356 | team = 394) -> DEGRADATION`.
You can reproduce this locally by using JMeter and `setup/performance-toolkit/benchmark.jmx` scenario.

**Comment by rossmitchell** [comment]
Hi @lenaorobei

Thanks for taking a look at this. I'm not sure why this would cause a performance degradation, but I will look into it.

I've not really used JMeter before and I'm having some difficulties running the performance tests locally. I've followed the instructions here https://github.com/magento/magento2/tree/2.3/setup/performance-toolkit and searched through the github issues to try and get these running.

Based on what I've found I put together a script here https://gist.github.com/rossmitchell/8c8df1dc632aa0c9d8a7cb8928882eb8 to setup a magento instance with the sample data and then run the scenario, however I'm not sure how to get the changes compared to the mainline branch.

I'm also seeing failures in the output saying `text expected to contain /You deleted the category./` which when I look at the access logs look to be coming from these requests `"POST /admin/catalog/category/delete/id/${admin_category_id}/ HTTP/1.1"`

Can I ask you to have a quick look at the script that I'm using and see if there is anything obviously wrong with it?

Thanks

**Review by mslabko — APPROVED** [review]

**Comment by lenaorobei** [comment]
@rossmitchell we are investigating performance failure issue. I will keep you updated.

**Comment by lenaorobei** [comment]
Internal PAT build https://bamboo.corp.magento.com/browse/MCCE24-MCCE24569-1

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. Install Magento;
2. Copy _setup/performance-toolkit/profiles/ce/small.xml_ to a new file;
3. Update the file to have 2,000 simple products, 2,000 categories, and (500, 1000, 5000, 25000) catalog price rules;
4. Generate fixtures using the file;
5. Run the Catalog Rule Product index;

 _**With 500 catalog price rules**_

_Before:_
before500


_After:_
after500

The value has improved by **33%**

 _**With 1000 catalog price rules**_

_Before:_
before1000

_After:_
after1000

The value has improved by **67%**

 _**With 5000 catalog price rules**_

_Before:_
before5000


_After:_
after5000


 _**With 25000 catalog price rules**_

_Before:_
The indexing time was more than 2 hours

_After:_
after25000

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by lenaorobei** [comment]
As per discussion with @mslabko - perf degradation is expected.
Neds to be approved internally.

**Review by engcom-Charlie — CHANGES_REQUESTED** [review]
Hi @rossmitchell, could You fix failing tests?

**Comment by engcom-Charlie** [comment]
Hi @rossmitchell, I'm closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for your collaboration.

**Comment by torhoehn** [comment]
@lenaorobei Since #25889 still is an open issue, should this PR be really closed? The author already mentioned that he need some guidance. What's the state of the internal approval you mentioned before?

**Comment by rossmitchell** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

@sidolov, can we re-run the performance tests? Do we still get the performance degradation? If so - please explain in which case?

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Integration Tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Integration Tests

**Comment by torhoehn** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Integration Tests

**Comment by engcom-Dash** [comment]
Hi @rossmitchell 

Thanks for your contribution!
I am looking into the build failures.

Thanks

**Comment by engcom-Dash** [comment]
@magento run all tests

**Comment by engcom-Dash** [comment]
@magento run all tests

**Comment by engcom-Dash** [comment]
@magento run all tests

**Comment by engcom-Dash** [comment]
Following functional and integration test case are failing 
Functional Tests B2B
    AdminCreateCatalogPriceRuleToPercentTest
    AdminCreateCatalogPriceRuleByPercentTest
    AdminCreateCatalogPriceRuleToFixedTest
    AdminEnableAttributeIsUndefinedCatalogPriceRuleTest

Functional Tests CE
    StorefrontConfigurableProductOutOfStockCustomStockTest
    StorefrontGalleryConfigurableProductWithVisualSwatchAttributePrependMediaTest

Integration Tests
   testCustomerGroupExcludeWebsite

All these testcases failure is not related to this PR changes as checked.

**Review by engcom-Charlie — DISMISSED** [review]

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

Tested the PR before and after changes as per [above comments](https://github.com/magento/magento2/pull/25908#issuecomment-633413636).
After more updates re-tested now using same above steps and PR changes successfully resolved the issue as shown in the below screenshot.

Screenshot 2022-01-05 at 9 39 19 PM

Based on the [above comment](https://github.com/magento/magento2/pull/25908#issuecomment-1005476989) on auto-test failure analysis, moving this PR for _Merge In Progress_.

**Review by adifucan — APPROVED** [review]

**Comment by engcom-Alfa** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests

**Review by adifucan — CHANGES_REQUESTED** [review]
This PR breaks some functional scenarios (see functional tests) + integration test

File: app/code/Magento/CatalogRule/Model/Indexer/IndexBuilder.php [hunk 1]
```diff
@@ -309,9 +309,9 @@ protected function doReindexByIds($ids)
     {
         $this->cleanProductIndex($ids);
 
-        /** @var Rule[] $activeRules */
-        $activeRules = $this->getActiveRules()->getItems();
-        foreach ($activeRules as $rule) {
+        $activeRules = $this->getActiveRules();
+        while ($rule = $activeRules->fetchItem()) {
```
Comment (kandy) [inline app/code/Magento/CatalogRule/Model/Indexer/IndexBuilder.php:313]: Looks like using fetchItem make an improvement due to the fact that it does not call \Magento\CatalogRule\Model\ResourceModel\Rule\Collection::_afterLoad and it breaks functionality that requires data is loaded in this method

**Comment by engcom-Hotel** [comment]
Hello @rossmitchell,

Thanks for the contribution!

We request you to please resolve the conflicts of this PR so that we can proceed further with this.

Thanks

**Comment by engcom-Hotel** [comment]
Moved back to `On Hold` as conflicts needs to be resolve yet.

**Comment by engcom-Hotel** [comment]
Hello @rossmitchell,

As per this comment https://github.com/magento/magento2/pull/25908#issuecomment-1584334058, we are unable to move further on this PR until the conflicts have been resolved. 

Hence we are closing this PR for now. Please feel free to re-open it whenever you are ready or ask us for the same, we will be happy to re-open it.

Thanks

**Pull request closed without merging**